src/providers/Web3Provider.tsx

```diff
@@ -0,0 +1,277 @@
+  ethProvider: null,
+  sapphireEthProvider: null,
+  wRoseContractAddress: null,
+  wRoseContract: null,
+  account: null,
+  explorerBaseUrl: null,
+  networkName: null,
+}
+
+export const Web3Context = createContext<Web3ProviderContext>({} as Web3ProviderContext)
+
+export const Web3ContextProvider: FC<PropsWithChildren> = ({ children }) => {
+  const [state, setState] = useState<Web3ProviderState>({
+    ...web3ProviderInitialState,
+  })
+
+  const _init = async (account: string) => {
+    try {
+      const ethProvider = new ethers.providers.Web3Provider(window.ethereum)
+      const sapphireEthProvider = sapphire.wrap(ethProvider) as (ethers.providers.Web3Provider & sapphire.SapphireAnnex)
+
+      const network = await sapphireEthProvider.getNetwork()
+
+      if (!(network.chainId in NETWORKS)) {
+        return Promise.reject(new UnknownNetworkError('Unknown network!'))
+      }
+
+      const { wRoseContractAddress, explorerBaseUrl, networkName } = NETWORKS[network.chainId]
+
+      const wRoseContract = new ethers.Contract(
+        wRoseContractAddress,
+        WrappedRoseMetadata.output.abi,
+        sapphireEthProvider.getSigner(),
+      )
+
+      setState(prevState => ({
+        ...prevState,
+        isConnected: true,
+        ethProvider,
+        sapphireEthProvider,
+        wRoseContract,
+        account,
+        explorerBaseUrl,
+        networkName,
+        wRoseContractAddress
+      }))
+    } catch (ex) {
+      setState(prevState => ({
+        ...prevState,
+        isConnected: false,
+      }))
+
+      throw new Error('[Web3Context] Unable to initialize providers!')
+    }
+  }
+
+  const getBalance = async () => {
+    const { account, sapphireEthProvider } = state
+
+    if (!account || !sapphireEthProvider) {
+      throw new Error('[Web3Context] Unable to fetch balance!')
+    }
+
+    return await sapphireEthProvider.getBalance(account)
+    // return utils.formatEther(balanceString)
+  }
+
+  const getBalanceOfWROSE = async () => {
+    const { account, wRoseContract } = state
+
+    if (!account || !wRoseContract) {
+      throw new Error('[Web3Context] Unable to fetch WROSE balance!')
+    }
+
+    return await wRoseContract.balanceOf(account)
+  }
+
+  const isMetaMaskInstalled = async () => {
+    const provider = await detectEthereumProvider();
+
+    return !!window.ethereum && provider === window.ethereum;
+  }
+
+  const connectWallet = async () => {
+    const [account] = await window.ethereum.request?.({ method: 'eth_requestAccounts' })
+
+    if (!account) {
+      throw new Error('[Web3Context] Request account failed!')
+    }
+
+    await _init(account)
+  }
+
+  const _addNetwork = async (chainId: number) => {
+    if (chainId === 0x5afe) {
+
+      // Default to Sapphire Mainnet
+      await window.ethereum.request?.({
+        method: 'wallet_addEthereumChain',
+        params: [
+          {
+            chainId: '0x5afe',
+            chainName: 'Sapphire Mainnet',
```

Comment:
```suggestion
            chainName: 'Oasis Sapphire',
```
This doesn't match chainlist data, so metamask shows a warning
(also doesn't match our docs https://github.com/oasisprotocol/docs/blob/c7a21a48ed5d1fcf360e341be2ef4499c071a319/docs/dapp/sapphire/README.mdx#L49-L85 )

Reply:
Nice catch 👍

Reply:
New explorer URL should probably be fine, right?

Reply:
yeah